exporter/kafkaexporter/README.md

```diff
@@ -40,6 +40,7 @@ The following settings can be optionally configured:
   - The following encodings are valid *only* for **logs**.
     - `raw`: if the log record body is a byte array, it is sent as is. Otherwise, it is serialized to JSON. Resource and record attributes are discarded.
 - `partition_traces_by_id` (default = false): configures the exporter to include the trace ID as the message key in trace messages sent to kafka. *Please note:* this setting does not have any effect on Jaeger encoding exporters since Jaeger exporters include trace ID as the message key by default.
+- `partition_metrics_by_resource_attributes` (default = false) configures the exporter to include the hash of sorted resource attributes as the message partitioning key in metric messages sent to kafka.
```

Comment:
Is there a reason the key has to be the hashed value of all resource attributes?

I'm wondering if we could get the same functionality if we simply had users define which resource attribute they want to use as the key. In my mind it would be more straightforward and configurable for users, as well as simpler to implement and maintain internally.

Reply:
Hi @crobert-1 and thank you for the feedback!

In our time series db we do not store original data but instead we aggregate those into second/minute buckets. Thus, if all metrics are distributed across all Kafka partitions we end up each consumer having all series, causing O(Kafka partitions) memory overhead. To solve this we need to send metrics with same labels to be delivered to only single Kafka partition. Our initial idea was to calculate partition key based on all resource, scope and datapoint attribute - it would help to achieve the most even metrics distribution but would cause too high overhead on collector also most likely completely remove any batching. So we decided to partition by just resources. 

Regarding the specific attributes - since we are not 100% standardized on the resource attributes list across different language-sdks, I would prefer to have an option to calculate the partition key planely on all resources. This would help to avoid case of someone breaking the whole pipeline renaming sone key attribute(or multiple ones) identifying the instance, for example from `host.name` to `service.instance.id`. As well it would allow to not change configs in case of updates in semantic conventions.

I agree - for the sake of flexibility we can support partitioning only by specific attributes, leaving an option to partition by everything. I can change the config to following:

```
PartitionMetricsByResourceAttributes:
    Enabled: True
    AttributeKeys: 
        - key1
        - key2
```
Enabled is `False` by default
AttributeKeys is `empty` by default
If AttributeKeys is empty, all resources are used
If AttributeKeys are not empty, only specified keys are used.

Please let me know if you find this solution acceptable.

Reply:
> I would prefer to have an option to calculate the partition key planely on all resources. This would help to avoid case of someone breaking the whole pipeline renaming sone key attribute(or multiple ones) identifying the instance, for example from host.name to service.instance.id.

Wouldn't this mean the key changes if **any** of the resource attributes change, making the key even more likely to change? (Correct me if I'm wrong)

I don't have a strong preference against this solution, just trying to understand the pros and cons here 👍

Reply:
Yes, that's exactly the goal I want to achieve. In case of new attribute added, all linked metrics will rebalance causing x2 memory overhead, but only for the balancing period. This is perfectly acceptable for us.

Reply:
Works for me, thanks for clarifying!

Reply:
I've added the functionality to partition by specific keys, please have a look